torch/_custom_op.py

```diff
@@ -0,0 +1,292 @@
+            >>>     return tensor.new_empty(output_shape)
+            >>>
+            >>> x = torch.randn(2, 3, device='meta')
+            >>> y = custom_sum(x, 1)
+            >>> assert y.shape == (2,)
+
+        """
+
+        def inner(f):
+            library.impl(self._lib, self._opname, "Meta")(f)
+            return f
+
+        return inner
+
+
+def process_namespaced_schema_str(
+    namespaced_schema: str,
+) -> Tuple[str, str, FunctionSchema]:
+    ns, schema_str = parse_namespace(namespaced_schema)
+    schema = FunctionSchema.parse(schema_str)
+
+    if ns in RESERVED_NS:
+        raise ValueError(
+            f"{ns} is a reserved namespace, please choose something else. "
+            f"Found when trying to create a new CustomOp with schema "
+            f"{namespaced_schema}"
+        )
+
+    # Coming in the future. Requires us to have correct logic for
+    # the ADInplaceOrView key
+    if schema.kind() != SchemaKind.functional:
+        raise NotImplementedError(
+            f"NYI: CustomOp.define does not support non-functional function schema. Got: {schema}"
+        )
+
+    rets = schema.returns
+    is_non_mutating_view = len(rets) > 0 and any(
+        r.annotation is not None and not r.annotation.is_write for r in rets
+    )
+    if is_non_mutating_view:
+        raise NotImplementedError(
+            f"NYI: CustomOp.define does not support view functions. Got: {schema}"
+        )
+
+    # Requires us to have handling for factory functions
+    if not schema.arguments.has_tensor_arg():
+        raise NotImplementedError(
+            f"NYI: CustomOp.define does not support function schema with no Tensor inputs. Got: {schema}"
+        )
+    # Just seems weird so banning for now
+    if not schema.returns:
+        raise NotImplementedError(
+            f"NYI: CustomOp.define does not support function schema with no outputs. Got: {schema}"
+        )
+    return ns, schema_str, schema
+
+
+def parse_namespace(namespaced_entity: str) -> Tuple[str, str]:
+    names = namespaced_entity.split("::", 1)
+    if len(names) != 2:
+        raise ValueError(f"Expected there to be a namespace in {namespaced_entity}.")
+    return names[0], names[1]
+
+
+def validate_device_type(device_type: str) -> None:
+    if device_type not in SUPPORTED_DEVICE_TYPES:
+        raise ValueError(
+            f"CustomOp.impl(device_type={device_type}): we only support device_type "
+            f"in {SUPPORTED_DEVICE_TYPES}."
+        )
+
+
+def get_autograd_not_implemented_kernel(custom_op) -> Callable:
+    def autograd_not_implemented(*args, **kwargs) -> None:
+        if pytree.tree_any(
+            lambda x: isinstance(x, torch.Tensor) and x.requires_grad, (args, kwargs)
+        ):
+            raise RuntimeError("Autograd has not been implemented for operator")
+        # TODO(rzou): RAII guard should be contextmanager, or else bad things will happen
+        guard = _C._AutoDispatchBelowAutograd()
```

Comment:
probably a dumb q: what bad things can happen if we just use a try/finally block here?

Reply:
The object would not get deleted if we accidentally created a reference cycle with it, but if we're careful when writing the code then we should be fine. So this is more of an idiomatic thing where python encourages people to use context managers instead of rely on RAII, especially because other implementations of python (not cpython) do not need to guarantee this.